alchemist/alchemist-incarnation-biochemistry/src/main/java/it/unibo/alchemist/model/implementations/conditions/BiomolPresentInEnv.java

```diff
@@ -8,28 +8,106 @@
  */
 package it.unibo.alchemist.model.implementations.conditions;
 
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.apache.commons.math3.util.CombinatoricsUtils;
+import org.apache.commons.math3.util.FastMath;
+
 import it.unibo.alchemist.model.implementations.molecules.Biomolecule;
-import it.unibo.alchemist.model.implementations.nodes.EnvironmentNode;
+import it.unibo.alchemist.model.interfaces.Context;
+import it.unibo.alchemist.model.interfaces.Environment;
+import it.unibo.alchemist.model.interfaces.EnvironmentNode;
+import it.unibo.alchemist.model.interfaces.Node;
 /**
  *
  */
 public class BiomolPresentInEnv extends GenericMoleculePresent<Double> {
 
-    private static final long serialVersionUID = 6527297022655890961L;
+    private static final long serialVersionUID = 1L;
+
+    private final Environment<Double> environment;
 
     /**
+     * Initialize condition for extra-cellular environment, implemented as a set
+     * of {@link EnvironmentNode}.
      * 
-     * @param biomol 
-     * @param concentration 
-     * @param node 
+     * @param biomol
+     *            the {@link Biomolecule} which the condition is about.
+     * @param conc
+     *            the requested concentration.
+     * @param node
+     *            the node where this condition is located;
+     * @param env
+     *            the {@link Environment} where the node is located.
      */
-    public BiomolPresentInEnv(final Biomolecule biomol, final Double concentration, final EnvironmentNode node) {
-        super(biomol, node, concentration);
+    public BiomolPresentInEnv(final Environment<Double> env, final Node<Double> node, final Biomolecule biomol, final Double conc) {
+        super(node, biomol, conc);
+        environment = env;
     }
 
     @Override
-    public EnvironmentNode getNode() {
-        return (EnvironmentNode) super.getNode();
+    public double getPropensityConditioning() {
+        double quantityInEnvNodes = 0;
+        if (!getEnviromentNodesSurrounding().isEmpty()) {
+            quantityInEnvNodes = getEnviromentNodesSurrounding().stream()
+                    .parallel()
+                    .mapToDouble(n -> n.getConcentration(getBiomolecule()))
+                    .sum();
+        }
+        double quantityInLayers = 0;
+        if (environment.getLayer(getBiomolecule()).isPresent()) {
+            quantityInLayers = environment.getLayer(getBiomolecule()).get().getValue(environment.getPosition(getNode()));
```

Comment:
This method chain is unclear. You should not call the same stuff over and over:
```java
final Optional<Layer<Double>> layer = environment.getLayer(getBiomolecule())
```
and then use it. In general, avoid calling the same methods over and over.